teps/0120-canceling-concurrent-pipelineruns.md

```diff
@@ -102,7 +102,326 @@ queueing, or more advanced concurrency strategies.
+    bindings:
+    - name: pull-request-id
+      value: $(body.check_suite.pull_requests[0].id)
+    - name: head-sha
+      value: $(body.check_suite.head_sha)
+    concurrency:
+      params:
+      - name: pull-request-id
+      key: $(params.pull-request-id)
+      strategy: cancel
+    interceptors:
+      ref:
+        kind: ClusterInterceptor
+        name: github
+    template:
+      spec:
+        params:
+        - name: head-sha
+        resourcetemplates:
+        - apiVersion: tekton.dev/v1beta1
+          kind: PipelineRun
+          spec:
+            pipelineRef:
+              name: ci-pipeline
+            params:
+            - name: head-sha
+              value: $(tt.params.head-sha)
+```
+
+TODO(Lee): Do we want to support multiple concurrency controls per Trigger? No clear use case for multiple cancels,
+but might want a cancel and a queue when that's supported later
+
+### Components of Concurrency Spec
+
+#### Params
+
+Concurrency parameter names should match the names of the TriggerBindings, and support substitution in the same way as parameters in TriggerTemplates.
+
+#### Key
+
+A string used to match PipelineRuns to each other. PipelineRuns created by the same Trigger with the same concurrency key are considered part of the same
+concurrency "group". We should support parameter substitution, and may choose to support substitution of context-related variables
+like [those supported in Pipelines](https://tekton.dev/docs/pipelines/variables/) (for example, `context.pipelineRun.namespace`).
+
+#### Strategy
+
+For the initial implementation of this proposal, we will support only canceling PipelineRuns.
+
+Initially supported strategies are:
+- "Cancel": [Cancel](https://tekton.dev/docs/pipelines/pipelineruns/#cancelling-a-pipelinerun) any running PipelineRuns with the same key.
+(When we support canceling TaskRuns, this will be the only supported cancelation strategy.)
+- "CancelRunFinally": [Gracefully cancel](https://tekton.dev/docs/pipelines/pipelineruns/#gracefully-cancelling-a-pipelinerun) any running PipelineRuns with the same key.
+The new PipelineRun will start without waiting for the existing PipelineRuns' finally Tasks to complete.
+- "StopRunFinally": [Gracefully stop](https://tekton.dev/docs/pipelines/pipelineruns/#gracefully-cancelling-a-pipelinerun) any running PipelineRuns with the same key.
+The new PipelineRun will start without waiting for the existing PipelineRuns' finally Tasks to complete.
+
+### Reconciler logic
+
+When a Trigger with a concurrency spec creates a new PipelineRun, it will substitute the parameters in the concurrency key and apply the concurrency key as a label
+with key "triggers.tekton.dev/concurrency". It will use an informer to find all PipelineRuns with the same concurrency key from the same Trigger, using label selectors.
+The reconciler will patch any matching PipelineRuns as canceled before creating the new PipelineRun, but will not wait for cancelation to complete.
```

Comment:
From my point of view, the only "job" of triggers should be to create that label and *most likely* create the `PipelineRun` as pending, and have *something else* handle that label (aka not the triggers controller).